packages/component-base/src/focus-utils.js

```diff
@@ -0,0 +1,202 @@
+/**
+ * @license
+ * Copyright (c) 2021 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+
+/**
+ * Returns true if the element is hidden with `visibility: hidden` or `display: none`, false otherwise.
+ *
+ * @param {!HTMLElement} element
+ * @return {boolean}
+ */
+function isElementHidden(element) {
+  // Check inline style first to save a re-flow. If looks good, check also
+  // computed style.
+  let style = element.style;
+  if (style.visibility === 'hidden' || style.display === 'none') {
+    return true;
+  }
+  style = window.getComputedStyle(element);
+  return style.visibility === 'hidden' || style.display === 'none';
+}
+
+/**
+ * Returns true if the element is focusable, otherwise false.
+ *
+ * @param {!HTMLElement} element
+ * @return {boolean}
+ */
+function isElementFocusable(element) {
+  // From http://stackoverflow.com/a/1600194/4228703:
+  // There isn't a definite list, it's up to the browser. The only
+  // standard we have is DOM Level 2 HTML
+  // https://www.w3.org/TR/DOM-Level-2-HTML/html.html, according to which the
+  // only elements that have a focus() method are HTMLInputElement,
+  // HTMLSelectElement, HTMLTextAreaElement and HTMLAnchorElement. This
+  // notably omits HTMLButtonElement and HTMLAreaElement. Referring to these
+  // tests with tabbables in different browsers
+  // http://allyjs.io/data-tables/focusable.html
+
+  // Elements that cannot be focused if they have [disabled] attribute.
+  if (element.matches('input, select, textarea, button, object')) {
+    return element.matches(':not([disabled])');
+  }
+  // Elements that can be focused even if they have [disabled] attribute.
+  return element.matches('a[href], area[href], iframe, [tabindex], [contentEditable]');
```

Comment:
Should there be more checks regarding accessibility? Like aria-hidden? See e.g. tabindex inside aria-hidden https://www.w3.org/WAI/standards-guidelines/act/rules/aria-hidden-no-focusable-content-6cfa84/ for more details with positive and negative test examples

Reply:
About this particular PR, I would try to keep the implementation within its original shape as much as possible because the main idea here is more about decomposing and moving functions to the `component-base` package.

Although, regarding the `aria-hidden` attribute, does it need special care? According to the first link you shared, the attribute doesn't make the element not focusable but rather prevents its announcement. Am I right?

BTW, I extracted these changes into a [separate PR](https://github.com/vaadin/web-components/pull/3141), let's continue the discussion there if you want.